test/jasmine/assets/check_transitions.js

```diff
@@ -0,0 +1,115 @@
+        });
+    }, Promise.resolve([]));
+}
+
+function clockTick(currentNow, milliseconds) {
+    Date.now = function() {
+        return currentNow + milliseconds;
+    };
+    d3.timer.flush();
+}
+
+// Using the methodology from http://eng.wealthfront.com/2017/10/26/testing-d3-transitions/
+module.exports = function checkTransition(gd, mock, animateOpts, transitionOpts, tests) {
+    if(!transitionOpts) {
+        transitionOpts = {
+            transition: {
+                duration: 500,
+                easing: 'linear'
+            },
+            frame: {
+                duration: 500
+            }
+        };
+    }
+    // Prepare chain
+    var now = Date.now;
+    var startTime;
+    var currentTime;
+    var p = [
+        function() {
+            return Plotly.newPlot(gd, mock)
+            .then(function() {
+                // Check initial states if present
+                for(var i = 0; i < tests.length; i++) {
+                    if(tests[i][0] === 0) assert(tests[i]);
+                }
+            });
+        },
+        function() {
+            // Hijack Date.now
+            startTime = Date.now();
+            currentTime = 0;
+            clockTick(startTime, 0);
+            Plotly.animate(gd, animateOpts, transitionOpts);
+            return Promise.resolve(true);
+        }
+    ];
+
+    var checkTests = tests.map(function(test) {
+        return function() {
+            if(test[0] === 0) return Promise.resolve(true);
+            if(test[0] !== currentTime) {
+                clockTick(startTime, test[0]);
+                currentTime = test[0];
+            }
+            return assert(test);
+        };
+    });
+
+    // Run all tasks
+    return promiseSerial(p.concat(checkTests))
+        .then(function() {
+            Date.now = now;
+        });
+};
+
+// A test array is made of
+// [ms since start of transition, selector, (attr|style), name of attribute, array of values to be found]
+// Ex.: [0, '.point path', 'style', 'fill', ['rgb(31, 119, 180)', 'rgb(31, 119, 180)', 'rgb(31, 119, 180)']]
+function assert(test) {
+    var msg = 'at ' + test[0] + 'ms, selection ' + test[1] + ' has ' + test[3];
+    var cur = [];
+    Plotly.d3.selectAll(test[1]).each(function(d, i) {
+        if(test[2] === 'style') cur[i] = this.style[test[3]];
+        if(test[2] === 'attr') cur[i] = Plotly.d3.select(this).attr(test[3]);
+    });
+    switch(test[3]) {
+        case 'd':
+            assertAttr(cur, test[4], round, msg);
+            break;
+        case 'transform':
+            assertAttr(cur, test[4], round, msg);
+            break;
+        default:
+            assertAttr(cur, test[4], Lib.identity, msg);
+    }
+    return Promise.resolve(true);
+}
+
+function assertAttr(A, B, cb, msg) {
+    var a = cb(A);
+    var b = cb(B);
+    expect(a).withContext(msg + ' equal to ' + JSON.stringify(b)).toEqual(b);
+}
+
+function round(str) {
```

Comment:
I guess this is fine.

I would have prefered extracting the numbers out of the transform strings and then use [`.toBeCloseArray`](https://github.com/plotly/plotly.js/blob/46866fdbb1fdf2e12437a48252f190d07807aa00/test/jasmine/assets/custom_matchers.js#L58-L73), but this wrapper is a huge :trophy:

Reply:
**This comment is non-blocking**

Reply:
> I would have prefered extracting the numbers out of the transform strings and then use [`.toBeCloseArray`](https://github.com/plotly/plotly.js/blob/46866fdbb1fdf2e12437a48252f190d07807aa00/test/jasmine/assets/custom_matchers.js#L58-L73), but this wrapper is a huge

This might indeed be necessary because of the discrepancy in the size of text elements in different browsers/OS which ends affecting the exact transform strings.

Reply:
Hmm, looks like there are indeed cross-browser differences.

Running `npm run test-jasmine -- bar` on Chrome 70 in Ubuntu 18.04, the two cases with on-graph bar text fail for me, see details below:

<details>
Chrome 70.0.3538 (Linux 0.0.0) bar tweening for vertical bar height and text position FAILED
	Error: at 0ms, selection text.bartext has transform equal to ["translate(90 58)","translate(269 239)","translate(450 210)"]: Expected $[0] = 'translate(90 58)' to equal 'translate(90 56)'.
	Expected $[1] = 'translate(269 239)' to equal 'translate(270 237)'.
	Expected $[2] = 'translate(450 210)' to equal 'translate(450 208)'.
	    at <Jasmine>
	    at assertAttr (/tmp/test/jasmine/assets/check_transitions.js:109:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210918:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:98:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210907:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:48:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210857:43
	Error: at 300ms, selection text.bartext has transform equal to ["translate(90,166)","translate(269,130)","translate(450,187)"]: Expected $[0] = 'translate(90,166)' to equal 'translate(90,164)'.
	Expected $[1] = 'translate(269,130)' to equal 'translate(270,128)'.
	Expected $[2] = 'translate(450,187)' to equal 'translate(450,185)'.
	    at <Jasmine>
	    at assertAttr (/tmp/test/jasmine/assets/check_transitions.js:109:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210918:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:98:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210907:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:69:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210878:20
	Error: at 600ms, selection text.bartext has transform equal to ["translate(90 239)","translate(269 58)","translate(450 172)"]: Expected $[0] = 'translate(90 239)' to equal 'translate(90 237)'.
	Expected $[1] = 'translate(269 58)' to equal 'translate(270 56)'.
	Expected $[2] = 'translate(450 172)' to equal 'translate(450 170)'.
	    at <Jasmine>
	    at assertAttr (/tmp/test/jasmine/assets/check_transitions.js:109:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210918:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:98:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210907:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:69:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210878:20
Chrome 70.0.3538 (Linux 0.0.0) bar tweening for horizontal bar length and text position FAILED
	Error: at 0ms, selection text.bartext has transform equal to ["translate(99 230)","translate(505 140)","translate(248 50)"]: Expected $[0] = 'translate(99 230)' to equal 'translate(100 229)'.
	Expected $[1] = 'translate(505 140)' to equal 'translate(506 139)'.
	Expected $[2] = 'translate(248 50)' to equal 'translate(249 49)'.
	    at <Jasmine>
	    at assertAttr (/tmp/test/jasmine/assets/check_transitions.js:109:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210918:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:98:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210907:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:48:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210857:43
	Error: at 300ms, selection text.bartext has transform equal to ["translate(227,230)","translate(389,140)","translate(287,50)"]: Expected $[0] = 'translate(227,230)' to equal 'translate(228,229)'.
	Expected $[1] = 'translate(389,140)' to equal 'translate(391,139)'.
	Expected $[2] = 'translate(287,50)' to equal 'translate(287,49)'.
	    at <Jasmine>
	    at assertAttr (/tmp/test/jasmine/assets/check_transitions.js:109:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210918:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:98:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210907:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:69:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210878:20
	Error: at 600ms, selection text.bartext has transform equal to ["translate(312 230)","translate(313 140)","translate(312 50)"]: Expected $[0] = 'translate(312 230)' to equal 'translate(314 229)'.
	Expected $[1] = 'translate(313 140)' to equal 'translate(314 139)'.
	Expected $[2] = 'translate(312 50)' to equal 'translate(313 49)'.
	    at <Jasmine>
	    at assertAttr (/tmp/test/jasmine/assets/check_transitions.js:109:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210918:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:98:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210907:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:69:1 <- /tmp/1ea018d52a3b10a9a939f0f793fee4b4.browserify.js:210878:20
Chrome 70.0.3538 (Linux 0.0.0): Executed 112 of 112 (2 FAILED) (4.036 secs / 3.5 secs)
</details>

Reply:
I introduced tolerance for `transform` checks in https://github.com/plotly/plotly.js/pull/4180/commits/86ba54ad454c804c29298017b7a497df0829c828

Reply:
One of those tests is still failing for me:

<details>
Chrome 70.0.3538 (Linux 0.0.0) bar tweening for vertical bar height and text position FAILED
	Error: at 0ms, selection text.bartext has transform equal to ["translate(89.796875 58)","translate(269.375 238.5)","translate(449.78125 210)"]: Expected 89.796875,58,269.375,238.5,449.78125,210 to be close to 90,56,270,236.5,450,208 
	    at <Jasmine>
	    at assertCloseTo (/tmp/test/jasmine/assets/check_transitions.js:117:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210926:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:99:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210908:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:48:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210857:43
	Error: at 300ms, selection text.bartext has transform equal to ["translate(89.796875,166.3)","translate(269.375,130.2)","translate(449.78125,187.2)"]: Expected 89.796875,166.3,269.375,130.2,449.78125,187.2 to be close to 90,164.3,270,128.20000000000002,450,185.2 
	    at <Jasmine>
	    at assertCloseTo (/tmp/test/jasmine/assets/check_transitions.js:117:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210926:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:99:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210908:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:69:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210878:20
	Error: at 600ms, selection text.bartext has transform equal to ["translate(89.796875 238.5)","translate(269.375 58)","translate(449.78125 172)"]: Expected 89.796875,238.5,269.375,58,449.78125,172 to be close to 90,236.5,270,56,450,170 
	    at <Jasmine>
	    at assertCloseTo (/tmp/test/jasmine/assets/check_transitions.js:117:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210926:67)
	    at assert (/tmp/test/jasmine/assets/check_transitions.js:99:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210908:13)
	    at /tmp/test/jasmine/assets/check_transitions.js:69:1 <- /tmp/dbdeaeadcc8a079807751cc57fad1a9c.browserify.js:210878:20
</details>

----

Bumping the tolerance to `2.5` (from `2`) make the tests pass.

Some info for future PRs: in my experience precision-based tolerances (e.g. `- Math.log10(2 * precision)` doesn't work so well for cross-browser differences in text sizing/positioning (like here). Absolute-value-based tolerances (e.g. using `toBeWithin` or `toBeWithinArray`) appears to work better.

Reply:
I set the tolerance to 3 in 16c88f8!

Reply:
Very very nice! Thanks!!